### PR TITLE
Centralize route authentication

### PR DESCRIPTION
**Pull request by heyitsmeuralex**
Enables authentication with *any route* via:
* sessionID in post body
* sessionID in query string
* X-Session-ID in request headers

Before, some routes would expect sessionID to be in the post body, some in the query string, etc. This makes them not care. Documentation has been updated to match.

No frontend changes have been made, but it should be trivial to evict all `?sessionID` and `{ sessionID }` declarations in api.js calls. Just using X-Session-ID is definitely much easier.

---

@towerofnix I'd recommend some extra testing for this particular PR than usual. I don't normally work on the backend all that much, and this is an API-wide change. I may have overlooked something.

@TheInitializer this may make weecent's code simpler or something, so you might want to take a look once/if this gets merged.

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -1,16 +1,18 @@
...
+The project's API is publically available to anybody with access to the actual server on which it is hosted. [HTTP endpoints](#http-endpoints) provide virtually all methods of interaction from the client towards the server, while [WebSocket events](#websocket-events) let the server send messages to the client. Common data received from the server (such as users or messages) always follows [particular formats](#objects). It would be wise to understand and expect [authorization](#authorization) to be required. All endpoints, unless s General information which doesn't particularly fit anywhere else can be found in the appendix-esque section [Etc](#etc), and any questions one might have can be posted to the [issue tracker](https://github.com/towerofnix/decent/issues).
```
Comment (towerofnix) [inline docs/api.md]: `All endpoints, unless s` Oops?

File: api.js [hunk 2]
```diff
@@ -1194,11 +1184,6 @@ module.exports = async function attachAPI(app, {wss, db}) {
 
   app.get('/api/user-list', [
     ...middleware.runIfCondition(() => shouldUseAuthorization, [
-      ...middleware.loadVarFromQuery('sessionID', false),
-      ...middleware.runIfVarExists('sessionID',
-        middleware.getSessionUserFromID('sessionID', 'sessionUser')
-      ),
-
```
Comment (towerofnix) [inline api.js]: This gets rid of the `sessionUser` middleware variable so it is probably broken.

File: api.js [hunk 3]
```diff
@@ -1375,8 +1359,6 @@ module.exports = async function attachAPI(app, {wss, db}) {
   ])
 
   app.post('/api/delete-sessions', [
-    ...middleware.loadVarFromBody('sessionIDs'),
-
```
Comment (towerofnix) [inline api.js]: This should not be removed - `sessionIDs`, not `sessionID`.

File: docs/api.md [hunk 4]
```diff
@@ -325,6 +327,15 @@ Note that some endpoints do not require authorization:
...
 * [`GET /api/session/:sessionID`](#get-apisessionsessionid)
 
+### Authentication:
+
+When a request is made to the API, the server searches for a session ID given in the request using:
+* `sessionID` in POST data
+* `?sessionID` in query string
+* `X-Session-ID` header
+
+If the server requires [authorization](#authorization) and the session ID could not be found or pointed to an unauthenticated user, the request is immediately terminated with a 403 and error message. We recommend simply sending the `X-Session-ID` header with _all_ requests, even on non-authorization-requiring servers.
```
Comment (towerofnix) [inline docs/api.md]: I don't like "we" - this should be `it's recommended to` 😃

File: api.js (hunk 2 above)
Comment (heyitsmeuralex) [reply api.js]: As far as I know all authenticated endpoints get the sessionUser variable for free.

Comment (towerofnix) [reply api.js]: I mean, the only solution to this argument is to *test* `/api/user-list` (but I'm not quite awake yet). I don't see anything that's *added* to save `sessionUser`, though; only that it now saves <code>session<em>Id</em></code> for free.

File: docs/api.md (hunk 4 above)
Comment (heyitsmeuralex) [reply docs/api.md]: `It's probably simpler to send` thanks

File: docs/api.md [hunk 5]
```diff
@@ -1,6 +1,6 @@
...
+The project's API is publically available to anybody with access to the actual server on which it is hosted. [HTTP endpoints](#http-endpoints) provide virtually all methods of interaction from the client towards the server, while [WebSocket events](#websocket-events) let the server send messages to the client. Common data received from the server (such as users or messages) always follows [particular formats](#objects). It would be wise to understand and expect [authorization](#authorization) to be required. All endpoints, unless General information which doesn't particularly fit anywhere else can be found in the appendix-esque section [Etc](#etc), and any questions one might have can be posted to the [issue tracker](https://github.com/towerofnix/decent/issues).
```
Comment (towerofnix) [inline docs/api.md]: Still not fixed - "All endpoints, unless General information which doesn't particularly fit anywhere else can be found in the appendix-esque section..."

"General information..." is a separate sentence from "All endpoints, unless". (All endpoints unless should be removed or completed.)

Comment (heyitsmeuralex) [reply docs/api.md]: whooops

**Pull request merged**